### PR TITLE
Added new getUsersByIds( ) API

### LINKED ISSUE #27235: API to search users by a list of ids
Some apps like https://github.com/owncloud/customgroups/issues/23 have a list of user ids (ex: group members) and need a way to get the display name of all said users.

One approach would be to join directly on the database level but since we're crossing app boundaries, that's probably not a good idea.

So here the idea is to add a new API to the user manager:
`IUserManager->getUsersByIds($userIds)`

It takes an array of user ids and asks every backend for all the given user ids.
On the DB backend that would be `SELECT * FROM oc_users where uid in (?)`, replace "?" with the list of user ids, but paginate / chunk.

Once the user account table from https://github.com/owncloud/core/issues/23558 is implemented, then it the method could directly retrieve the info from the accounts table.

@DeepDiver1975 @butonic @jvillafanez does that make sense ?

This is a blocker for https://github.com/owncloud/customgroups/issues/23 and I don't want to implement the logic from above in a hacky way inside the app.

### PR DESCRIPTION
**Pull request by noveens**
## Description

Fixes #27235 
I have added a new API by the name getUsersByIds( ) under the interface IUserManager.

## Related Issue

https://github.com/owncloud/core/issues/27235

## Motivation and Context

Adds a new API to give a list of displaynames based on input array of user IDs.

## How Has This Been Tested?

I have not tested this, any guidance so as to how to test is highly appreciated :-)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Remaining work:
- [ ] Update the Documentation.
- [ ] Write new tests for this API

### CODE REVIEW

**Comment by noveens** [comment]
@PVince81 ,
can you review?

File: lib/private/User/Manager.php [hunk 1]
```diff
@@ -467,4 +468,33 @@ public function getByEmail($email) {
 			return $this->get($uid);
 		}, $userIds);
 	}
+	/**
+	 * @param string[] $userIds
+	 * @return string[] displayname of respective uid
```
Comment (PVince81) [inline lib/private/User/Manager.php]: Please return `IUser` instances. See how the other functions do it.

**Review by PVince81 — COMMENTED** [review]
It's a good start. But the user manager needs to return an array of `IUser` objects (which is the `User` instance).

Those objects already contain the display name and might even contain the fallback in case display name is null.

File: lib/private/User/Manager.php [hunk 2]
```diff
@@ -467,4 +468,33 @@ public function getByEmail($email) {
...
+	 * @since 10.0.0
+	 */
+	public function getUsersByIds($userIds) {
+		$queryBuilder = \OC::$server->getDatabaseConnection()->getQueryBuilder();
+		$displayNames = [];
+
+		foreach ($userIds as $uid) {
```
Comment (PVince81) [inline lib/private/User/Manager.php]: This is going to be expensive.

Please use a `select ... where uid in (...)` to retrieve several users at once.

Comment (noveens) [inline lib/private/User/Manager.php]: @PVince81 ,
that would give me a list of the displaynames but how would I know that which uid has which displayname?

**Comment by PVince81** [comment]
Arghh, I just realized that `new User()` doesn't even contain the display name https://github.com/owncloud/core/blob/master/lib/private/User/Manager.php#L161

and calling `getDisplayName` would do an extra SQL query just to get it: https://github.com/owncloud/core/blob/master/lib/private/User/User.php#L123

The general problem is that the User Manager is just a frontend to many possible user backends (DB, LDAP, IMAP, etc). Maybe you noticed that every method here is iterating over every backend and call the same function. You can't directly read the needed information from the DB because it might not even be in the DB (ex: LDAP).

You need to iterate over every backend and call the function `getUsersByIds()`. Since this method doesn't exist, you need to create it there. The code you put here, move it to the database backend: https://github.com/owncloud/core/blob/master/lib/private/User/Database.php#L60

Then you might need to add a new optional argument to the `new User()` constructor to be able to accept a display name, if provided. If not provided, let the old logic get it as it did in the past.

There might be more to do, but let's first do what I mentioned until here.

**Comment by noveens** [comment]
@PVince81 ,
I guess then that I should leave working on this PR and continue once #27165 is merged.
This would make the database structure clearer and also easier to test.

File: lib/private/User/Manager.php (hunk 2 above)
Comment (PVince81) [inline lib/private/User/Manager.php]: by selecting both "uid" and "displayname"

**Comment by PVince81** [comment]
Okay, so the user account table has been merged and released with 10.0 beta.

It should now be possible to adjust this PR to query the account table directly.

**Comment by PVince81** [comment]
actually need to go through the `$this->accountMapper` so might need a new method there as well

**Comment by PVince81** [comment]
I think we won't need this after all, or not very soon.

Also it might need a different approach anyway.

Closing.

**Pull request closed without merging**